### PR TITLE
Added two classes. One create, delete and otherwise manage services. Another to make running php as a service easy.

### PR DESCRIPTION
**Pull request by credomane**
These are two classes I made 4-5 years ago. Spent some time today updating them and thought I'd send them as a PR.

I just noticed the examples folder contains a similar and combined version of my version here. 
Sending this PR to bounce around ideas and compare with the version in examples.

### CODE REVIEW

File: phplib/Pecl/Win32Service/WindowsService.php [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+        $this->setState(WIN32_SERVICE_STOP_PENDING);
+        $this->serviceStop();
+        $this->setState(WIN32_SERVICE_STOPPED);
+    }
+
+    /**
+     * Preforms any setup that is needed by ServiceTick().
+     * Should return quickly.
+     */
+    protected function serviceStart() {
```
Comment (macintoshplus) [inline phplib/Pecl/Win32Service/WindowsService.php:69]: Use abstract function instead function ?

File: phplib/Pecl/Win32Service/WindowsService.php [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+    }
+
+    /**
+     * Called repeatedly but no more than once per second.
+     * Should return as often as possible or make arrangements
+     * to manually call serviceResponder().
+     */
+    protected function serviceTick() {
```
Comment (macintoshplus) [inline phplib/Pecl/Win32Service/WindowsService.php:78]: Use abstract function instead function ?

File: phplib/Pecl/Win32Service/WindowsService.php [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+    }
+
+    /**
+     * Performs any clean up before the service exits.
+     */
+    protected function serviceStop() {
```
Comment (macintoshplus) [inline phplib/Pecl/Win32Service/WindowsService.php:85]: Use abstract function instead function ?

File: phplib/Pecl/Win32Service/WindowsService.php [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+namespace Pecl\Win32Service;
+
+/**
+ * Override functions as needed by your application.
+ * serviceStart(), serviceTick(), and serviceStop() are designed to be overridden.
+ * Call serviceMain() to start your service. Function will not return until windows stops the service.
+ * Optionally, override serviceMain() for 100% control.
+ */
+class WindowsService {
```
Comment (macintoshplus) [inline phplib/Pecl/Win32Service/WindowsService.php:19]: Use abstract class instead class ?

**Comment by macintoshplus** [comment]
Thanks for your contribution.
Can you move your classes into the exemple folder ?
Or why not integrate your classes into the ext ? In C ?

**Review by macintoshplus — COMMENTED** [review]
Thanks for your contribution.
Can you move your classes into the exemple folder ?
Or why not integrate your classes into the ext ? In C ?

**Comment by credomane** [comment]
> Can you move your classes into the example folder ?

Dang it. I messed up my local git by doing everything on master (not even the up-to-date php7 branch). No clue what I was thinking. The three PRs I made were gonna end up an ugly mess if I didn't start over before pushing them to github. Guess I stuck these in the wrong folder the second time around.

> why not integrate your classes into the ext ? In C ?

I would love to do that but I haven't a clue how to actually make classes inside the C ext itself. If you have a guide or tutorial somewhere to get me started I'll more than gladly do that and would honestly prefer that route. Not knowing how to is the whole reason these php files exists as they do. :P

> Use abstract class instead?

Making it an abstract class would be proper sense you are meant to extend this class with your own.

> Use abstract functions instead?

I'm torn on how to handle this. If someone decided to override serviceMain() in a subclass for full control they potentially would not need to override serviceStart(), serviceTick(), or serviceStop() at all. Any suggestions how to handle that case?

**Comment by macintoshplus** [comment]
> I would love to do that but I haven't a clue how to actually make classes inside the C ext itself. If you have a guide or tutorial somewhere to get me started I'll more than gladly do that and would honestly prefer that route. Not knowing how to is the whole reason these php files exists as they do. :P

For write extension, I use this book written in french : https://leanpub.com/developper-une-extension-php
I have read the PHP Manual : http://php.net/manual/en/internals2.structure.php
And, I reade the other ext, or I find a tutorial on Internet.

> Making it an abstract class would be proper sense you are meant to extend this class with your own.

Use Interface and Trait is not a good answer ?

> I'm torn on how to handle this. If someone decided to override serviceMain() in a subclass for full control they potentially would not need to override serviceStart(), serviceTick(), or serviceStop() at all. Any suggestions how to handle that case?

IMHO, if you want full control write your class from scratch ! no ? This code is an exemple not a required way.

**Comment by macintoshplus** [comment]
Your idea has rewritten into the service-library available here: https://github.com/win32service/service-library

Thanks for your help.

**Pull request closed without merging**